edb/server/protocol/auth_ext/webauthn.py

```diff
@@ -230,3 +234,207 @@ async def _delete_registration_challenges(
                 "user_handle": user_handle,
             },
         )
+
+    async def create_authentication_options_for_email(
+        self,
+        *,
+        webauthn_provider: config.WebAuthnProvider,
+        email: str,
+    ) -> Tuple[str, bytes]:
+        # Find credential IDs by email
+        result = await execute.parse_execute_json(
+            self.db,
+            """
+select ext::auth::WebAuthnFactor {
+    user_handle,
+    credential_id,
+}
+filter .email = <str>$email;""",
+            variables={
+                "email": email,
+            },
+            cached_globally=True,
+        )
+        result_json = json.loads(result.decode())
+        user_handles: set[str] = {x["user_handle"] for x in result_json}
+        assert len(user_handles) == 1
```

Comment:
Why do we assert that there's only a single user_handle associated with a given email? I thought we wanted to allow you to use the same email with multiple authenticators: https://github.com/edgedb/edgedb/pull/6791#discussion_r1483149382
Unless they should all have the same user_handle (which I guess was the reason for dropping the exclusive constraint on `(.email, .user_handle)`?), in which case in `create_registration_options_for_email` we should check if there's already a `WebAuthnFactor` for that email and re-use the `user_handle`. I don't know if there's some security issue with being able to retrieve an email's user_handle like that though?

Reply:
Yeah, we're _supposed_ to use the same user_handle if the user is "the same user" to allow multiple WebAuthn factors for a single user. You're right that I'm missing that check from the `create_registration_options_for_email` but I need to think about the UX here a bit more.

The ostensible reason for allowing multiple factors is to allow what are basically back-up factors (or multi-factors) so you can seamlessly move between passkeys, hardware authenticators, etc. but in our "registration" flow it's not super obvious that you might have multiple identities with the same email which point to a single `User` in the downstream application because we kind of leave that up to the application developer to do.

But, yeah I think at the very least we should reuse the same user handle if you provide an existing email, will fix.

Reply:
@jaclarke ⚡ [`ebda3e8` (#6792)](https://github.com/edgedb/edgedb/pull/6792/commits/ebda3e809602670e77b623b401cd6951aff914bf)

Reply:
From testing it seems there's another issue with this assert. If the email isn't registered, then there's no user-handles,the assert fails and we return an ise. I'm not sure what the correct error should be here? I know we're trying not to leak the existence of emails, but I don't think we can get around that as we can only return a challenge if we have the stored credentials for an email.

Reply:
Yeah, good catch: assert is the wrong tool here, we need to raise a specific exception for proper error handling.

As far as data leakage is concerned, I think we talked about `register` being fine to leak whether an email is signed up or not, but `authenticate` would always pretend it succeeds. From that perspective, I think we can return a sensible error here.

Reply:
@jaclarke ⚡ https://github.com/edgedb/edgedb/pull/6792/files/40d636291fca7bb5db36ebd142e7e58fa2214bdd..1431063a019694d9028966b1ae1cbce2d8c6e93b

Turns out I needed a bit more work to ensure the invariant that multiple factors with the same `email` have the same `user_handle`. Added a test for the new trigger.